File: bff/flows.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type tokenResponse struct {
	AccessToken  string          `json:"access_token,omitempty"`
	TokenType    string          `json:"token_type,omitempty"`
	ExpiresIn    int             `json:"expires_in,omitempty"`
	RefreshToken string          `json:"refresh_token,omitempty"`
	IDToken      string          `json:"id_token,omitempty"`
	Scope        string          `json:"scope,omitempty"`
	Raw          json.RawMessage `json:"raw,omitempty"`
}

type oidcError struct {
	Status int
	Body   string
	Code   string
}

func (e *oidcError) Error() string {
	if e.Code != "" {
		return fmt.Sprintf("token endpoint returned %d (%s): %s", e.Status, e.Code, truncate(e.Body, 200))
	}
	return fmt.Sprintf("token endpoint returned %d: %s", e.Status, truncate(e.Body, 200))
}

func exchangeCode(ctx context.Context, sess *session, code string) (*tokenResponse, error) {
	form := url.Values{
		"grant_type":   {"authorization_code"},
		"code":         {code},
		"redirect_uri": {sess.RedirectURI},
	}
	if sess.CodeVerifier != "" {
		form.Set("code_verifier", sess.CodeVerifier)
	}
	return tokenRequest(ctx, sess.Discovery.TokenEndpoint, form, sess.ClientID, sess.ClientSecret)
}

func clientCredentials(ctx context.Context, sess *session) (*tokenResponse, error) {
	form := url.Values{"grant_type": {"client_credentials"}}
	if len(sess.Scopes) > 0 {
		form.Set("scope", strings.Join(sess.Scopes, " "))
	}
	return tokenRequest(ctx, sess.Discovery.TokenEndpoint, form, sess.ClientID, sess.ClientSecret)
}

func refreshTokens(ctx context.Context, sess *session) (*tokenResponse, error) {
	if sess.Tokens == nil || sess.Tokens.RefreshToken == "" {
		return nil, fmt.Errorf("no refresh token available")
	}
	form := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {sess.Tokens.RefreshToken},
	}
	return tokenRequest(ctx, sess.Discovery.TokenEndpoint, form, sess.ClientID, sess.ClientSecret)
}

func tokenRequest(ctx context.Context, endpoint string, form url.Values, clientID, clientSecret string) (*tokenResponse, error) {
	if endpoint == "" {
		return nil, fmt.Errorf("no token_endpoint advertised")
	}
	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.SetBasicAuth(url.QueryEscape(clientID), url.QueryEscape(clientSecret))

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		var errResp struct {
			Error string `json:"error"`
		}
		_ = json.Unmarshal(body, &errResp)
		return nil, &oidcError{Status: resp.StatusCode, Body: string(body), Code: errResp.Error}
	}
	tr := &tokenResponse{}
	if err := json.Unmarshal(body, tr); err != nil {
		return nil, err
	}
	tr.Raw = body
	return tr, nil
}

File: worker/src/oidc/flows.ts
import type { TokenResponse } from '@repartee/shared/contract'
import type { SessionData } from '../session/sessionDO'

const TOKEN_REQUEST_TIMEOUT_MS = 10_000

export class OidcError extends Error {
  status: number
  body: string
  code?: string
  constructor(status: number, body: string, code?: string) {
    super(
      code
        ? `token endpoint returned ${status} (${code}): ${truncate(body, 200)}`
        : `token endpoint returned ${status}: ${truncate(body, 200)}`,
    )
    this.name = 'OidcError'
    this.status = status
    this.body = body
    this.code = code
  }
}

export interface TokenExchange {
  endpoint: string
  form: URLSearchParams
  clientId: string
  clientSecret: string
}

export async function exchangeCode(
  sess: SessionData,
  code: string,
  fetcher: typeof fetch = fetch,
): Promise<TokenResponse> {
  const form = new URLSearchParams({
    grant_type: 'authorization_code',
    code,
    redirect_uri: sess.redirectURI ?? '',
  })
  if (sess.codeVerifier) form.set('code_verifier', sess.codeVerifier)
  return tokenRequest(
    {
      endpoint: sess.discovery?.token_endpoint ?? '',
      form,
      clientId: sess.clientId ?? '',
      clientSecret: sess.clientSecret ?? '',
    },
    fetcher,
  )
}

export async function clientCredentials(
  sess: SessionData,
  fetcher: typeof fetch = fetch,
): Promise<TokenResponse> {
  const form = new URLSearchParams({ grant_type: 'client_credentials' })
  if (sess.scopes && sess.scopes.length > 0) form.set('scope', sess.scopes.join(' '))
  return tokenRequest(
    {
      endpoint: sess.discovery?.token_endpoint ?? '',
      form,
      clientId: sess.clientId ?? '',
      clientSecret: sess.clientSecret ?? '',
    },
    fetcher,
  )
}

export async function refreshTokens(
  sess: SessionData,
  fetcher: typeof fetch = fetch,
): Promise<TokenResponse> {
  const refresh = sess.tokens?.refresh_token
  if (!refresh) throw new Error('no refresh token available')
  const form = new URLSearchParams({ grant_type: 'refresh_token', refresh_token: refresh })
  return tokenRequest(
    {
      endpoint: sess.discovery?.token_endpoint ?? '',
      form,
      clientId: sess.clientId ?? '',
      clientSecret: sess.clientSecret ?? '',
    },
    fetcher,
  )
}

export async function tokenRequest(
  { endpoint, form, clientId, clientSecret }: TokenExchange,
  fetcher: typeof fetch = fetch,
): Promise<TokenResponse> {
  if (!endpoint) throw new Error('no token_endpoint advertised')

  const ac = new AbortController()
  const timer = setTimeout(() => ac.abort(), TOKEN_REQUEST_TIMEOUT_MS)
  let resp: Response
  try {
    resp = await fetcher(endpoint, {
      method: 'POST',
      headers: {
        'content-type': 'application/x-www-form-urlencoded',
        accept: 'application/json',
        // RFC 7617: raw UTF-8 username:password, base64-encoded. The Go BFF
        // URL-escapes the credentials before encoding which is technically
        // wrong; this implementation deliberately does it correctly. See
        // §Risks #4 in the plan.
        authorization: `Basic ${btoa(`${clientId}:${clientSecret}`)}`,
      },
      body: form.toString(),
      signal: ac.signal,
    })
  } finally {
    clearTimeout(timer)
  }

  const body = await resp.text()
  if (resp.status !== 200) {
    let code: string | undefined
    try {
      const parsed = JSON.parse(body) as { error?: string }
      if (typeof parsed.error === 'string') code = parsed.error
    } catch {
      // upstream did not send JSON; surface the raw body via OidcError
    }
    throw new OidcError(resp.status, body, code)
  }

  let parsed: TokenResponse
  try {
    parsed = JSON.parse(body) as TokenResponse
  } catch (err) {
    throw new Error(`decoding token response: ${(err as Error).message}`)
  }
  parsed.raw = JSON.parse(body)
  return parsed
}

function truncate(s: string, n: number): string {
  return s.length <= n ? s : s.slice(0, n) + '…'
}
